File: 2914-find-the-safest-path-in-a-grid/find-the-safest-path-in-a-grid.cpp
class Solution {
public:
    int ROWS;
    int COLS; 
    const int directions[4][2] = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}}; 

    int pathContainsSf(vector<vector<int>>& grid, int sf) {
        std::queue<std::pair<int, int>> q;
        std::vector<vector<bool>> visited(ROWS, vector<bool>(COLS, false));
        q.push(make_pair(0, 0));
        visited[0][0] = true; 
        while (!q.empty()) {
            int s = q.size();
            for (int i = 0; i < s; ++i) {
                pair<int, int> cur = q.front();
                q.pop();
                if (cur.first == ROWS - 1 && cur.second == COLS - 1) {
                    return true; // valid path found; 
                }

                for (const auto& d : directions) {
                    int x = cur.first + d[0]; 
                    int y = cur.second + d[1];
                    if (
                        (x < 0 || y < 0) ||
                        (x >= ROWS || y >= COLS) ||
                        visited[x][y]
                    ) {
                        continue;
                    }
                    if (grid[x][y] >= sf) {
                        visited[x][y] = true; 
                        q.push(make_pair(x, y)); 
                    }
                }
            }
        }
        return false; 
    }
    int maximumSafenessFactor(vector<vector<int>>& grid) {
        std::queue<std::pair<int, int>> q; 
        std::vector<std::pair<int, int>> thiefs; 
        ROWS = grid.size(); 
        COLS = grid[0].size(); 
        for (int i = 0; i < ROWS; ++i) {
            for (int j = 0; j < COLS; ++j) {
                if (grid[i][j] == 1) {
                    pair<int, int> p = make_pair(i, j); 
                    q.push(p);
                    grid[i][j] = 0; // thieves
                } else {
                    grid[i][j] = -1;
                }
            }
        }
                
        int level = 0; 
        while (!q.empty()) {
            int s = q.size(); 
            for (int i = 0; i < s; ++i) {
                std::pair<int, int> cur = q.front();
                q.pop(); 
                for (const auto &d : directions) {
                    int x = cur.first + d[0]; 
                    int y = cur.second + d[1];
                    
                    if (
                        x < 0 || y < 0 || 
                        x >= ROWS || y >= COLS ||
                        grid[x][y] >= 0
                    ) {
                        continue; // visited
                    }
                    grid[x][y] = level + 1; // update safety factor
                    q.push(make_pair(x,  y));
                }
            }
            level++; 
        }
        
        int start = 0; 
        int end = grid[0][0]; // the starting point contains the max possible sf
        int res = -1; // start point gurantees thou
        while (start <= end) {
            int mid = (start + end) / 2;
            if (pathContainsSf(grid, mid)) {
                res = mid;
                start = mid + 1; // try to find a better path
            } else {
                end = mid - 1; // when mid becomes 0, path always contains it
            }
        }
        return res; 
    }
};


/*
Basically fnd all paths from starting -> said point
For each path you should have a safeness factor,
    If there exists a coordinate that is closer to a thief, you have
    to update the safety factor of that path
At the end of each path, select the path with the highest factor

How do you determine at every point the closeness to a thief? 
    It seems inefficent at every point we branch out to find a 1, that's 
    really just O(v + e)^2
    Maybe we can do a precompuation? Given all the thiefs, we store the min
    distance nearest to a thief. Then we replace all theiefs with 0 
    But if every cell is a thief that means we are doing lots of multiple of DFS
    BUT we can do a single BFS with all starting points of thiefs
    That's like spreading the waves, if the cell has been visited before (value > 0)
    the we can guarentee that's the minimum factor that we have seen 

/
After getting all the safeness factor of each cell. The starting point contains
the maximum possible safeness factor, cuz even if we have other cell that have
better safeness factor, the path want's to have the minimum of it. 
*/

File: 2914-find-the-safest-path-in-a-grid/find-the-safest-path-in-a-grid.java
class Solution {
    public int ROWS;
    public int COLS; 
    public int[][] directions = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}}; 
    public int maximumSafenessFactor(List<List<Integer>> grid) {
        ROWS = grid.size();
        COLS = grid.get(0).size(); 
        adjustGrid(grid); 
        return modifiedDijkstra(grid); 
    }
    // modify dijkstra to priortise paths with higher maximum safety value
    // since we want to track the minimum of the max path, before push the new coordinate
    // modify the coordinate to take the minimum of the path we have seen 
    public int modifiedDijkstra(List<List<Integer>> grid) {
        PriorityQueue<ArrayList<Integer>> maxHeap = new PriorityQueue<>(
            (a, b) -> b.get(0).compareTo(a.get(0))
        );
        ArrayList<Integer> startPoint = new ArrayList<>();
        startPoint.add(grid.get(0).get(0)); 
        startPoint.add(0); 
        startPoint.add(0);
        grid.get(0).set(0, -1); //visited
        maxHeap.add(startPoint); 

        while (maxHeap.size() != 0) {
            List<Integer> cur = maxHeap.poll(); 
            if (cur.get(1) == (ROWS - 1) && cur.get(2) == (COLS - 1)) {
                    return cur.get(0); // max safety factor
            }
            for (int[] d : directions) {
                int a = cur.get(1) + d[0];
                int b = cur.get(2) + d[1]; 

                if (!withinBounds(a, b) || grid.get(a).get(b) == -1) {
                    continue; 
                }
                int minPath = Math.min(cur.get(0), grid.get(a).get(b));
                ArrayList<Integer> neigh = new ArrayList<>();
                neigh.add(minPath);
                neigh.add(a);
                neigh.add(b); 
                grid.get(a).set(b, -1); // mark as visited
                maxHeap.add(neigh); 
            }
        }
        return -1; // won't reach here

    }
    public void adjustGrid(List<List<Integer>> grid) { 
        
        Queue<List<Integer>> startPoints = new LinkedList<>(); 
        for (int i = 0; i < ROWS; ++i) {
            for (int j = 0; j < COLS; ++j) {
                if (grid.get(i).get(j).equals(1)) {
                    ArrayList<Integer> coords = new ArrayList<>();
                    coords.add(i); 
                    coords.add(j); 
                    startPoints.add(coords); 
                    grid.get(i).set(j, 0); 
                } else {
                    grid.get(i).set(j, -1); 
                }
            }
        }
        int layer = 0; 
        while (!startPoints.isEmpty()) {
            int size = startPoints.size(); 
            for (int i = 0; i < size; ++i) {
                List<Integer> cur = startPoints.poll(); 
                for (int[] d : directions) {
                    int x = cur.get(0) + d[0];
                    int y = cur.get(1) + d[1]; 
                    if (!withinBounds(x, y) || grid.get(x).get(y) != -1) {
                        continue; 
                    }
                    grid.get(x).set(y, layer + 1); // mark as visited
                    List<Integer> newCoords = new ArrayList<>();
                    newCoords.add(x);
                    newCoords.add(y);
                    startPoints.add(newCoords); 
                }
            }
            layer += 1; 
        }
        return;
    }

    public boolean withinBounds(int x, int y) {
        return (x >= 0 && x < ROWS) && (y >= 0 && y < COLS); 
    }
}